httpclient/response_test.go

```diff
@@ -1,29 +1,120 @@
 package httpclient
 
 import (
+	"bytes"
 	"context"
 	"io"
 	"net/http"
 	"strings"
 	"testing"
 
+	"github.com/google/go-cmp/cmp"
 	"github.com/stretchr/testify/require"
 )
 
-func TestSimpleRequestRawResponse(t *testing.T) {
-	c := NewApiClient(ClientConfig{
+func make200Response(body string) *http.Response {
+	return &http.Response{
+		StatusCode: 200,
+		Body:       io.NopCloser(strings.NewReader(body)),
+	}
+}
+
+func mockClient(resp *http.Response) *ApiClient {
+	return NewApiClient(ClientConfig{
 		Transport: hc(func(r *http.Request) (*http.Response, error) {
-			return &http.Response{
-				StatusCode: 200,
-				Body:       io.NopCloser(strings.NewReader("Hello, world!")),
-				Request:    r,
-			}, nil
+			resp.Request = r
+			return resp, nil
 		}),
 	})
-	var raw []byte
-	err := c.Do(context.Background(), "GET", "/a", WithResponseUnmarshal(&raw))
-	require.NoError(t, err)
-	require.Equal(t, "Hello, world!", string(raw))
+}
+
+func TestWithResponseUnmarshal_structWithContent(t *testing.T) {
+	type structWithContents = struct {
+		Contents io.ReadCloser
+	}
+	want := structWithContents{
+		Contents: io.NopCloser(strings.NewReader("foo bar")),
+	}
+
+	var got structWithContents
+	c := mockClient(make200Response("foo bar"))
+	gotErr := c.Do(context.Background(), "GET", "/a", WithResponseUnmarshal(&got))
+
+	if gotErr != nil {
+		t.Errorf("WithResponseUnmarshal(): want no error, got: %s", gotErr)
+	}
```

Comment:
You can use the `require` and/or `assert` packages for a shorthand.

```suggestion
	gotErr := c.Do(context.Background(), "GET", "/a", WithResponseUnmarshal(&got))
	require.NoError(t, gotErr, "WithResponseUnmarshal")
```

Reply:
I tend to avoid using libraries like `testify` because they are not idiomatic and are arguably against the Databricks' Go style guidelines (see refs below).

IMO, testify can be quite convenient in some use cases but it adds cognitive load for the developers and the readers that are not familiar with the library (i.e. it is one additional mini-language to learn). Adding that cognitive load doesn't seem to be worth the couple of lines it removes.

Some references:

- https://google.github.io/styleguide/go/decisions#assertion-libraries (not in Databricks' go style)
- https://go.dev/wiki/CodeReviewComments#useful-test-failures

Reply:
From offline discussion: we decided to keep using `testify` as it is the most common practice in the Databrick's SDK and CLI.